### PR TITLE
A template for probe sub command development which need to handle hierarchical scenario

### PR DESCRIPTION
**Pull request by hu-weihua**
This is not a new probe sub command
This is just a template for probe sub command development which need to handle hierarchical scenario

### CODE REVIEW

File: xCAT-probe/subcmds/code_template [hunk 1]
```diff
@@ -0,0 +1,343 @@
...
+# If $outputtarget has input value, that means do_main_job is running on MN, so every message needed to print on STDOUT should be written into pipe $outputtarget.
+# If $outputtarget has no value, that means do_main_job is running on SN, all message just need to print on STDOUT
+# Recommand to use probe_utils->send_msg() to handle message you plan to print out
+#------------------------------------
+sub do_main_job {
+    my $outputtarget = shift;
+    $outputtarget = "stdout" if (!$outputtarget);
+    my $rst = 0;
+
+    #<#DO YOUR OWN CHECKING JOB1i#>
```
Comment (daniceexi) [inline xCAT-probe/subcmds/code_template]: typo ``JOB1i``

File: xCAT-probe/subcmds/code_template [hunk 2]
```diff
@@ -0,0 +1,343 @@
...
+    } else {
+        foreach my $pid (keys %runningpid) {
+            kill 'KILL', $pid if ($runningpid{$pid});
+        }
+    }
+    ++$try;
+
+    sleep 1;
+    foreach my $pid (keys %runningpid) {
+        $runningpid{$pid} = 0 if (waitpid($pid, WNOHANG));
```
Comment (daniceexi) [inline xCAT-probe/subcmds/code_template]: do we get into an infinity loop here? Especially without waiting among the loop

File: xCAT-probe/subcmds/code_template [hunk 3]
```diff
@@ -0,0 +1,343 @@
...
+
+#-------------------------------------
+# summary all jobs output to display
+#-------------------------------------
+$rst = summary_all_jobs_output();
+
+exit $rst;
+
+
+
```
Comment (daniceexi) [inline xCAT-probe/subcmds/code_template]: general comment, please add more comments for your code.

File: xCAT-probe/subcmds/code_template [hunk 4]
```diff
@@ -111,17 +138,25 @@ sub caclulate_dispatch_cmd {
...
             my $snnodemap = xCAT::ServiceNodeUtils->get_ServiceNode(\@nodes, "xcat", "MN");
+
+            #print Dumper $snnodemap;
             my %newsnnodemap;
             foreach my $sn (keys %$snnodemap) {
-                if (grep(/^$sn$/, @snlist)) {
+                if (grep(/^$sn$/, @snlist)) {   # the node just belong to one SN
                     push(@{ $newsnnodemap{$sn} }, @{ $snnodemap->{$sn} });
-                } else {
+                } elsif ($sn =~ /(\w+),.+/) { # the node belong to more than one SN, count it into first SN
```
Comment (daniceexi) [inline xCAT-probe/subcmds/code_template]: do we still need to check whether the $1 is a sn?

File: xCAT-probe/subcmds/code_template [hunk 5]
```diff
@@ -111,17 +138,25 @@ sub caclulate_dispatch_cmd {
...
+                    push(@{ $newsnnodemap{$1} }, @{ $snnodemap->{$sn} });
+                } else { # the nodes don't belong to any SN will be handled by MN
                     push(@{ $newsnnodemap{mn} }, @{ $snnodemap->{$sn} });
```
Comment (daniceexi) [inline xCAT-probe/subcmds/code_template]: do we need to check it really is MN?

File: xCAT-probe/subcmds/code_template [hunk 6]
```diff
@@ -215,7 +255,7 @@ pipe $pipe_parent_read, $pipe_child_write;
     #handle job in MN
     $mnjobpid = fork();
     if (!defined($mnjobpid)) {
-        probe_utils->send_msg("$output", "f", "fork process to handle MN job failed: $!");
+        probe_utils->send_msg("$output", "f", "fock process to handle MN job failed: $!");
```
Comment (daniceexi) [inline xCAT-probe/subcmds/code_template]: why change it to ``fock`` from ``fork``?

File: xCAT-probe/subcmds/code_template [hunk 7]
```diff
@@ -232,13 +272,13 @@ pipe $pipe_parent_read, $pipe_child_write;
...
     foreach my $sn (keys %dispatchcmd) {
-        my $snjobcmd = "xdsh $sn -s \"$dispatchcmd{$sn}\"";
+        my $snjobcmd = "xdsh $sn -s \"$dispatchcmd{$sn}\" 2>&1";
 
         #print "$sn = $snjobcmd\n";
         my $snjobfd;
         my $snjobpid;
         if (!($snjobpid = open($snjobfd, "$snjobcmd |"))) {
-            probe_utils->send_msg("$output", "f", "fork process to dispatch cmd $snjobcmd to $sn failed: $!");
+            probe_utils->send_msg("$output", "f", "fock process to dispatch cmd $snjobcmd to $sn failed: $!");
```
Comment (daniceexi) [inline xCAT-probe/subcmds/code_template]: ``fock``

**Comment by zet809** [comment]
No more comments, and it is an example to guide user to write their own script, so merge it.

**Pull request merged**